### PR TITLE
ENH: include assign_self_weight (formerly fill_diagonal) in Graph

### LINKED ISSUE #668: fill diagonal in graph
is still not yet implemented. The laziest way i can do this now is 

```python
g = Graph.build_kernel(gdf, bandwidth=1000)
sp = g.sparse
sp = sp.tolil()
sp.setdiag(1)
g = Graph.from_sparse(sp, ids=gdf.index)
```

which is what we do currently for W. Not sure if there's something more performant than roundtripping through sparse, given the new format? maybe some pandas-fu or something?

### PR DESCRIPTION
**Pull request by martinfleis**
Closes #668

@knaaptime this includes pandas-fu version as well as roundtrip via sparse for now. Timing below. 

```py
import geopandas as gpd
from geodatasets import get_path
from libpysal import graph

df = gpd.read_file(get_path("geoda.ncovr"))
contig  = graph.Graph.build_contiguity(df)

%%timeit
contig.fill_diagonal()
5.31 ms ± 108 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%%timeit
contig.fill_diagonal_sparse()
16.3 ms ± 92.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Here, the sparse version is about 3x slower. This ratio seems to be more or less consistent with any size of the graph.

### CODE REVIEW

File: libpysal/graph/base.py [hunk 1]
```diff
@@ -1419,6 +1421,45 @@ def eliminate_zeros(self):
...
+
+        Value for each ``focal == neighbor`` location in the graph is set to ``val``.
+
+        Parameters
+        ----------
+        val : float | array-like
+            Defines the value(s) to which the weights matrix diagonal should
+            be set. If a constant is passed then each element along the
+            diagonal will get this value (default is 1). An array of length
+            Graph.n can be passed to set explicit values to each element along
```
Comment (jGaboardi) [inline libpysal/graph/base.py:1435]:
```suggestion
            ``Graph.n`` can be passed to set explicit values to each element along
```

**Review by jGaboardi — COMMENTED** [review]
Looking good.

**Review by jGaboardi — APPROVED** [review]

File: libpysal/graph/base.py [hunk 2]
```diff
@@ -1419,6 +1421,45 @@ def eliminate_zeros(self):
...
+            name="weight",
+        )
+        adj = (
+            pd.concat([self.adjacency.drop(self.isolates), addition])
+            .reindex(self.unique_ids, level=0)
+            .reindex(self.unique_ids, level=1)
+        )
+        return Graph(adj, is_sorted=True)
+
+    def fill_diagonal_sparse(self):
```
Comment (knaaptime) [inline libpysal/graph/base.py:1457]: nice! not sure if we are keeping both? but this needs to accept a `val`

Comment (martinfleis) [reply libpysal/graph/base.py:1457]: I kept it here for an easy comparison. I will remove it :). Can't think of a reason to keep both.

Comment (knaaptime) [reply libpysal/graph/base.py:1457]: thats what i figured 🚀

**Comment by sjsrey** [comment]
https://github.com/pysal/libpysal/issues/668#issuecomment-1868027857

**Review by jGaboardi — COMMENTED** [review]
Will the terminology in this PR be affected by the discussion in #668?

**Comment by martinfleis** [comment]
> Will the terminology in this PR be affected by the discussion in #668?

Yes. I'll update this PR as soon we have an agreement.

**Comment by martinfleis** [comment]
Updated as per #668. Check the docstring please if it makes sense.

File: libpysal/graph/base.py [hunk 3]
```diff
@@ -1419,6 +1421,38 @@ def eliminate_zeros(self):
         zeros = (self._adjacency == 0) != isolates
         return Graph(self._adjacency[~zeros], is_sorted=True)
 
+    def assign_self_weight(self, weight=1):
+        """Assign values to edges represeting self-weight.
```
Comment (jGaboardi) [inline libpysal/graph/base.py:1425]:
```suggestion
        """Assign values to edges representing self-weight.
```

File: libpysal/graph/base.py [hunk 4]
```diff
@@ -1419,6 +1421,38 @@ def eliminate_zeros(self):
...
+
+        Value for each ``focal == neighbor`` location in the graph is set to ``weight``.
```
Comment (jGaboardi) [inline libpysal/graph/base.py:1427]:
```suggestion
        The value for each ``focal == neighbor`` location in the graph is set to ``weight``.
```

File: libpysal/graph/base.py [hunk 5]
```diff
@@ -1419,6 +1421,38 @@ def eliminate_zeros(self):
...
+
+        Parameters
+        ----------
+        weight : float | array-like
+            Defines the value(s) to which the weight of self-loops should be set. If a
```
Comment (jGaboardi) [inline libpysal/graph/base.py:1432]: Are we sticking with "self-loop" here or going with "self-weight?"

File: libpysal/graph/base.py [hunk 6]
```diff
@@ -1419,6 +1421,38 @@ def eliminate_zeros(self):
...
+            constant is passed then each self-loop will get this value (default is 1).
```
Comment (jGaboardi) [inline libpysal/graph/base.py:1433]: Same here.

File: libpysal/graph/base.py [hunk 7]
```diff
@@ -1419,6 +1421,38 @@ def eliminate_zeros(self):
...
+            An array of length ``Graph.n`` can be passed to set explicit values to each
+            self-lopp (assumed to be in the same order as original data).
```
Comment (jGaboardi) [inline libpysal/graph/base.py:1435]:
```suggestion
            self-loop (assumed to be in the same order as original data).
```

Same here.

File: libpysal/graph/base.py [hunk 8]
```diff
@@ -1419,6 +1421,38 @@ def eliminate_zeros(self):
...
+
+        Returns
+        -------
+        Graph
+            A new Graph with added self-weights
```
Comment (jGaboardi) [inline libpysal/graph/base.py:1440]:
```suggestion
            A new ``Graph`` with added self-weights.
```

**Review by knaaptime — APPROVED** [review]

File: libpysal/graph/base.py [hunk 9]
```diff
@@ -1419,6 +1421,39 @@ def eliminate_zeros(self):
         zeros = (self._adjacency == 0) != isolates
         return Graph(self._adjacency[~zeros], is_sorted=True)
 
+    def assign_self_weight(self, weight=1):
+        """Assign values to edges representing self-weight.
+
+        The value for each ``focal == neighbor`` location in the graph is set to ``weight``.
```
Comment (jGaboardi) [inline libpysal/graph/base.py:1427]: Looks like `pre-commit` is complaining that this line is too long.

```suggestion
        The value for each ``focal == neighbor`` location in
        the graph is set to ``weight``.
```

**Review by jGaboardi — APPROVED** [review]

**Comment by jGaboardi** [comment]
Shall we update the title of the issue to reflect changes in terminology?

**Review by sjsrey — APPROVED** [review]

**Pull request merged**